lib/uktest/test.c

```diff
@@ -0,0 +1,241 @@
+	unsigned int count = 0;
+	struct uk_testsuite *suite;
+	struct uk_testcase *testcase;
+
+	UK_TAILQ_FOREACH(suite, &uk_testsuite_list, next) {
+		uk_testsuite_for_each_case(suite, testcase) {
+			count += testcase->total_asserts;
+		}
+	}
+
+	return count;
+}
+
+unsigned int
+uk_test_assert_failed_count(void)
+{
+	unsigned int count = 0;
+	struct uk_testsuite *suite;
+	struct uk_testcase *testcase;
+
+	UK_TAILQ_FOREACH(suite, &uk_testsuite_list, next) {
+		uk_testsuite_for_each_case(suite, testcase) {
+			count += testcase->failed_asserts;
+		}
+	}
+
+	return count;
+}
+
+#ifdef CONFIG_LIBUKTEST_PRINT_STATS
+#include <uk/plat/console.h>
+
+#define UK_TEST_STATS_INIT_CLASS UK_INIT_CLASS_LATE
+#define UK_TEST_STATS_INIT_PRIO  9 /* As late as possible. */
+
+#if CONFIG_LIBUKDEBUG_ANSI_COLOR
+#define UK_TEST_STAT_FAILED	UK_ANSI_MOD_BOLD \
+				UK_ANSI_MOD_COLORFG(UK_ANSI_COLOR_RED)
+#else /* CONFIG_LIBUKDEBUG_ANSI_COLOR */
+#define UK_TEST_STAT_FAILED
+#endif /* !CONFIG_LIBUKDEBUG_ANSI_COLOR */
+
+static int
+uk_test_print_stats(void)
+{
+	int failed;
+
+	uk_printd("\nTest Summary:\n");
+
+	/*
+	 * Test suites
+	 */
+
+	uk_printd(" - Suites:     ");
+	failed = uk_testsuite_failed_count();
+	if (failed > 0)
+		uk_printd(UK_TEST_STAT_FAILED
+			  "%d failed"
+			  UK_ANSI_MOD_RESET
+			  ", ", failed);
+	uk_printd("%d total\n", uk_testsuite_count());
+
+	/*
+	 * Test cases
+	 */
+
+	uk_printd(" - Cases:      ");
+	failed = uk_testcase_failed_count();
+	if (failed > 0)
+		uk_printd(UK_TEST_STAT_FAILED
+			  "%d failed"
+			  UK_ANSI_MOD_RESET
+			  ", ", failed);
+	uk_printd("%d total\n", uk_testcase_count());
+
+	/*
+	 * Assertions
+	 */
+
+	uk_printd(" - Assertions: ");
+	failed = uk_test_assert_failed_count();
+	if (failed > 0)
+		uk_printd(UK_TEST_STAT_FAILED
+			  "%d failed"
+			  UK_ANSI_MOD_RESET
+			  ", ", failed);
+	uk_printd("%d total\n", uk_test_assert_count());
+
+	return 0;
+}
+uk_initcall_class_prio(
+	uk_test_print_stats,
+	UK_TEST_STATS_INIT_CLASS,
+	UK_TEST_STATS_INIT_PRIO
+);
+#endif /* CONFIG_LIBUKTEST_PRINT_STATS */
+
+int
+uk_testsuite_add(struct uk_testsuite *suite)
+{
+	UK_ASSERT(suite);
+	UK_TAILQ_INSERT_TAIL(&uk_testsuite_list, suite, next);
+
+	return testsuite_count++;
+}
+
+int
+uk_testsuite_run(struct uk_testsuite *suite)
+{
+	int ret = 0;
+	struct uk_testcase *testcase;
+
+	if (suite->init) {
+		ret = suite->init(suite);
+
+		if (ret != 0) {
+			uk_pr_err("Could not initialize test suite: %s",
+				  suite->name);
+			goto ERR_EXIT;
+		}
+	}
+
+	/* Reset the number of failed cases before running each case. */
+	suite->failed_cases = 0;
+
+	uk_testsuite_for_each_case(suite, testcase) {
+#ifdef CONFIG_LIBUKTEST_LOG_TESTS
+		uk_printd(LVLC_TESTNAME "test:" UK_ANSI_MOD_RESET
+			  " [%s:%s]\n",
+			  suite->name,
+			  testcase->name
+		);
+#endif /* CONFIG_LIBUKTEST_LOG_TESTS */
+
+		testcase->func(testcase);
+
+		/* If one case fails, the whole suite fails. */
+		if (testcase->failed_asserts > 0)
+			suite->failed_cases++;
```

Comment:
This logic is not correct; it increases the number of failed cases for every failed assertion.  It should only increase the failed case once a single assertion has failed.